// @shopify/shopify-app-session-storage/build/ts/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultRdbmsSessionStorageMigratorOptions = exports.defaultSessionStorageMigratorOptions = exports.InvalidMigrationConfigurationError = exports.MigrationOperation = void 0;
var tslib_1 = require("tslib");
/**
 * Defines what is needed for a migration to be execute
 */
var MigrationOperation = /** @class */ (function () {
    function MigrationOperation(migrationName, migrationFunction) {
        this.migrationName = migrationName;
        this.migrationFunction = migrationFunction;
    }
    return MigrationOperation;
}());
exports.MigrationOperation = MigrationOperation;
/**
 * When the configuration for migration is not consistent
 */
var InvalidMigrationConfigurationError = /** @class */ (function (_super) {
    tslib_1.__extends(InvalidMigrationConfigurationError, _super);
    function InvalidMigrationConfigurationError(message) {
        return _super.call(this, message) || this;
    }
    return InvalidMigrationConfigurationError;
}(Error));
exports.InvalidMigrationConfigurationError = InvalidMigrationConfigurationError;
exports.defaultSessionStorageMigratorOptions = {
    migrationDBIdentifier: 'shopify_sessions_migrations',
    migrations: [],
};
exports.defaultRdbmsSessionStorageMigratorOptions = {
    migrationDBIdentifier: 'shopify_sessions_migrations',
    migrationNameColumnName: 'migration_name',
    migrations: [],
};
